File: medium/901. Online Stock Span/901. Online Stock Span.py
class StockSpanner:

    def __init__(self):
        self.stack = []

    def next(self, price: int) -> int:
        if not self.stack or self.stack[-1][1] > price:
            self.stack.append([1, price])
            return 1
        else:
            span = 0
            while self.stack and self.stack[-1][1] <= price:
                span += self.stack.pop()[0]

            self.stack.append([span + 1, price])
            return span + 1


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: medium/901. Online Stock Span/901. Online Stock Span.go
package main

type StockSpanner struct {
	stack [][]int
}

func Constructor() StockSpanner {
	return StockSpanner{stack: [][]int{}}
}

func (this *StockSpanner) Next(price int) int {
	if len(this.stack) == 0 || this.stack[len(this.stack)-1][1] > price {
		this.stack = append(this.stack, []int{1, price})
		return 1
	} else {
		span := 0

		for len(this.stack) > 0 && this.stack[len(this.stack)-1][1] <= price {
			prevPrice := this.stack[len(this.stack)-1]
			this.stack = this.stack[:len(this.stack)-1]
			span += prevPrice[0]
		}

		this.stack = append(this.stack, []int{span + 1, price})

		return span + 1
	}
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */
